<!-- EVALUATE VISUAL ODOMETRY -->
<launch>

    <param name="use_sim_time" value="true"/>
    
    <!-- #################
            BAG PLAYER
         ################# 
     MUST HAVE
     /stereo_cam/left/image_raw
     /stereo_cam/left/camera_info
     /stereo_cam/right/image_raw
     /stereo_cam/right/camera_info
     -->
     <node pkg="rosbag" type="play" name="rosbag" args="-r 0.5 /home/anaritapereira/Desktop/experiencias/camiao/bagfiles/bags_scania2/scania_long.bag --clock"/>
  
    <arg name="calibrated" default="1" />
    <arg name="respawn" default="true" />
    <arg name="decimation_x" default="1"/>
    <arg name="decimation_y" default="1"/>  
    <arg name="camera_nodelet_manager_name" default="cam_manager" />
    
    <group ns="stereo/wide">

        <!-- ###################
                NODELET MANAGER
             ################### -->    
        <node pkg="nodelet" type="nodelet" name="$(arg camera_nodelet_manager_name)" args="manager" output="screen" respawn="$(arg respawn)" />

        <!-- #################
                DEBAYER
             ################# -->
        <node pkg="nodelet" type="nodelet" name="cam_debayer_left"
            args="load image_proc/debayer $(arg camera_nodelet_manager_name)" output="screen" respawn="$(arg respawn)" >
            <remap from="image_raw" to="left/image_raw"/>
            <remap from="image_mono" to="left/image_mono"/>
            <remap from="image_color" to="left/image_color"/>
        </node>

        <node pkg="nodelet" type="nodelet" name="cam_debayer_right"
            args="load image_proc/debayer $(arg camera_nodelet_manager_name)" output="screen" respawn="$(arg respawn)" >
            <remap from="image_raw" to="right/image_raw"/>
            <remap from="image_mono" to="right/image_mono"/>
            <remap from="image_color" to="right/image_color"/>
        </node>

        <!-- #################
                RECTIFY
             ################# -->
        <node pkg="nodelet" type="nodelet" name="cam_rectify_left"
            args="load image_proc/rectify $(arg camera_nodelet_manager_name)" output="screen" respawn="$(arg respawn)" >
            <remap from="image_mono" to="left/image_mono"/>
            <remap from="image_rect" to="left/image_rect"/>
        </node>

        <node pkg="nodelet" type="nodelet" name="cam_rectify_right"
            args="load image_proc/rectify $(arg camera_nodelet_manager_name)" output="screen" respawn="$(arg respawn)" >
            <remap from="image_mono" to="right/image_mono"/>
            <remap from="image_rect" to="right/image_rect"/>
        </node>

        <!-- #################
                CROP
             ################# -->
        <node pkg="nodelet" type="nodelet" name="cam_crop_left" 
            args="load image_proc/crop_decimate $(arg camera_nodelet_manager_name)" output="screen" respawn="$(arg respawn)" >
            <param name="decimation_x" type="int" value="$(arg decimation_x)" />
            <param name="decimation_y" type="int" value="$(arg decimation_y)" />
            <param name="x_offset" type="int" value="40" />
            <param name="y_offset" type="int" value="30" />
            <param name="width" type="int" value="560" />
            <param name="height" type="int" value="420" />
            <remap from="camera/image_raw" to="left/image_rect"/>
            <remap from="camera/image_info" to="left/camera_info"/>
            <remap from="camera_out/image_raw" to="/stereo_image/left/image_rect"/>
            <remap from="camera_out/image_info" to="/stereo_image/left/camera_info"/>
        </node>

        <node pkg="nodelet" type="nodelet" name="cam_crop_right" 
            args="load image_proc/crop_decimate $(arg camera_nodelet_manager_name)" output="screen" respawn="$(arg respawn)" >
            <param name="decimation_x" type="int" value="$(arg decimation_x)" />
            <param name="decimation_y" type="int" value="$(arg decimation_y)" />        
            <param name="x_offset" type="int" value="40" />
            <param name="y_offset" type="int" value="30" />
            <param name="width" type="int" value="560" />
            <param name="height" type="int" value="420" />
            <remap from="camera/image_raw" to="right/image_rect"/>
            <remap from="camera/image_info" to="right/camera_info"/>
            <remap from="camera_out/image_raw" to="/stereo_image/right/image_rect"/>
            <remap from="camera_out/image_info" to="/stereo_image/right/camera_info"/>
        </node>    
    
    </group>
    
    
    <!-- ###################
            Transformations
         ################### -->

    <node pkg="tf2_ros" type="static_transform_publisher" name="imu_link" args="-1.31 0 1.20 0 0 0 1 base_link imu" />
    
    <node pkg="tf2_ros" type="static_transform_publisher" name="cam_link" args="-0.30 0 1.49 -1.57 0 -1.67  base_link camera" />
    
    <node pkg="tf2_ros" type="static_transform_publisher" name="vodom_link" args="0 0 0 0 0 0 1  base_link vodom" />

    
    <!-- #################
            Viso2
         ################# -->
    <node pkg="viso2_ros" type="stereo_odometer" name="viso_odometer" output="screen" respawn="$(arg respawn)" >
        <remap from="stereo" to="stereo/wide"/>
        <remap from="image" to="image_rect"/>
        <param name="base_link_frame_id" value="base_link"/>
        <param name="odom_frame_id" value="vodom"/>
        <param name="sensor_frame_id" value="camera"/>
        <param name="bucket_width" value="1280"/>
        <param name="bucket_height" value="960"/>
        <param name="max_features" value="1000"/>
    </node> 
    
    <!-- #############################################
            GET VISUAL ODOMETRY AND GROUND TRUTH DATA
         #############################################-->
    <node pkg="evaluate_vo" type="get_vogt_node" name="get_data" args="/home/anaritapereira/evaluate_vo_plots odometry pose" output="screen">
        <remap from="get_data/visual_odometry" to="viso_odometer/odometry" />
        <remap from="get_data/ground_truth" to="carina2/currentPose" />
    </node>

</launch>
